worker/worker.go

```diff
@@ -1174,6 +1174,147 @@ func (w *worker) objectsHandlerPUT(jc jape.Context) {
+		return
+	} else if contractset != "" {
+		up.ContractSet = contractset
+	}
+
+	// attach gouging checker to the context
+	ctx = WithGougingChecker(ctx, up.GougingParams)
+
+	// attach contract spending recorder to the context.
+	ctx = WithContractSpendingRecorder(ctx, w.contractSpendingRecorder)
+
+	w.pool.setCurrentHeight(up.CurrentHeight)
+	usedContracts := make(map[types.PublicKey]types.FileContractID)
+
+	// fetch contracts
+	contracts, err := w.bus.Contracts(ctx)
+	if jc.Check("couldn't fetch contracts from bus", err) != nil {
+		return
+	}
+
+	// randomize order of contracts so we don't always upload to the same hosts
+	frand.Shuffle(len(contracts), func(i, j int) { contracts[i], contracts[j] = contracts[j], contracts[i] })
+
+	// keep track of slow hosts so we can avoid them in consecutive slab uploads
+	slow := make(map[types.PublicKey]int)
+
+	// generate a unique key for each object.
+	keys := make([]object.EncryptionKey, len(req))
+	var crs []io.Reader
+	for i, r := range req {
+		keys[i] = object.GenerateEncryptionKey()
+		crs = append(crs, keys[i].Encrypt(io.LimitReader(requestBody, int64(r.Length))))
+	}
+	cr := io.MultiReader(crs...)
+
+	var objectSize int
+	var slabs []object.Slab
+	for {
+		var s object.Slab
+		var length int
+		var slowHosts []int
+
+		lr := io.LimitReader(cr, int64(rs.MinShards)*rhpv2.SectorSize)
+		// move slow hosts to the back of the array
+		sort.SliceStable(contracts, func(i, j int) bool {
+			return slow[contracts[i].HostKey] < slow[contracts[j].HostKey]
+		})
+
+		// upload the slab
+		start := time.Now()
+		s, length, slowHosts, err = uploadSlab(ctx, w, lr, uint8(rs.MinShards), uint8(rs.TotalShards), contracts, w, w.uploadSectorTimeout, w.uploadMaxOverdrive, w.logger)
+		for _, h := range slowHosts {
+			slow[contracts[h].HostKey]++
+		}
+		if err == io.EOF {
+			break
+		} else if jc.Check(fmt.Sprintf("uploading slab failed after %v", time.Since(start)), err); err != nil {
+			w.logger.Errorf("couldn't upload data, err: %v", err)
+			return
+		}
+
+		objectSize += length
+		slabs = append(slabs, s)
+
+		for _, ss := range s.Shards {
+			if _, ok := usedContracts[ss.Host]; !ok {
+				for _, c := range contracts {
+					if c.HostKey == ss.Host {
+						usedContracts[ss.Host] = c.ID
+						break
+					}
+				}
+			}
+		}
+	}
+
+	// Make sure the amount of data that was uploaded matches the request.
+	var expectedObjectSize int
+	var lengths []int
+	for _, r := range req {
+		expectedObjectSize += int(r.Length)
+		lengths = append(lengths, int(r.Length))
+	}
+	if expectedObjectSize != objectSize {
+		jc.Error(fmt.Errorf("expected object size %v, got %v", expectedObjectSize, objectSize), http.StatusBadRequest)
+		return
+	}
+
+	// Split uploaded slabs into multiple objects and add them to the bus.
+	// TODO: f/u with new endpoint to atomically add a batch of objects.
+	splitObjects := object.SplitSlabs(slabs, lengths)
+	for i, slabs := range splitObjects {
```

Comment:
Would it be better to split this up to be more configurable? This endpoint could also hit redsolver's requirement of uploading the data separately from setting the object metadata in the bus. It would take only the raw data, like the existing object endpoint, then return something like:
```go 
type ObjectsUploadResponse struct {
	Slabs         []object.Slab                            `json:"slabs"`
	Key           object.EncryptionKey                     `json:"key"`
	UsedContracts map[types.PublicKey]types.FileContractID `json:"usedContracts"`
}
```

Then the bus could have an endpoint that takes
```go
// PackedObject contains the metadata for a packed object.
type PackedObject struct {
	Key    string `json:"key"`
	Length int    `json:"length"`
}

type ObjectsRequest struct {
	SlabsUploadResponse
	Objects       []PackedObject                           `json:"objects"`
}
```
It's an additional step, but splitting and adding would be handled atomically by the bus and you get rid of the weird request body.

Reply:
The tricky part is the encryption. Every object has its own key so that you can share a packed object without handing out access to the other files. Your `ObjectsUploadResponse` would need to return all the keys used for encryption which means the endpoint needs information about the slices. Alternatively the client can handle encryption but then it becomes increasingly hard to use it.

I do agree on the json at the beginning of the upload being a bit nasty but headers won't have the capacity to hold all the metadata for tightly packed objects.

Reply:
I think the `json` is definitely something we need to think about because it is quite nasty. I also agree headers put a serious constraint on the scalability of the feature as a whole.

What if we have a `PUT /objects/prepare` endpoint on the `bus` or `worker`? It would be consistent with other endpoints, we'd avoid the awkward json "prefix" and it might even be possible to reuse or extend the existing upload endpoint to be able to handle a prepared upload.